### PR TITLE
Create V2TweetLookupClient with Tweet fields

### PR DESCRIPTION
**Pull request by ConnorSinnott**
This PR adds partial support for the [Tweet Lookup](https://developer.twitter.com/en/docs/twitter-api/tweets/lookup/introduction) endpoints under Twitter's new V2 API. 

Twitter's new [Object Model](https://developer.twitter.com/en/docs/twitter-api/data-dictionary/object-model/tweet) consists of 5 objects:
- [Tweets](https://developer.twitter.com/en/docs/twitter-api/data-dictionary/object-model/tweet)
- [Users](https://developer.twitter.com/en/docs/twitter-api/data-dictionary/object-model/user)
- [Media](https://developer.twitter.com/en/docs/twitter-api/data-dictionary/object-model/media)
- [Polls](https://developer.twitter.com/en/docs/twitter-api/data-dictionary/object-model/media)
- [Places](https://developer.twitter.com/en/docs/twitter-api/data-dictionary/object-model/place)

In a GraphQL-esque style, a query for Tweets under a tweet endpoint can accept a series of "Expansions" and "Fields" to be materialized in the result. In this way, a query for recent tweets can include in its response fully materailized Users, Media, and Polls based on the fields requested.

This interlinked system makes the piece-by-piece introduction of the V2 model difficult, but it can be done by prohibiting requests for fields or expansions which reference data not yet modeled in the code.

### CODE REVIEW

File: src/test/scala/com/danielasfregola/twitter4s/http/clients/rest/v2/tweets/TwitterV2TweetLookupClientSpec.scala [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+          request.method === HttpMethods.GET
+          request.uri.endpoint === "https://api.twitter.com/2/tweets"
+          request.uri.rawQueryString === Some(
+            "expansions=attachments.poll_ids%2Cattachments.media_keys%2Cauthor_id%2Centities.mentions.username%2Cgeo.place_id%2Cin_reply_to_user_id%2Creferenced_tweets.id%2Creferenced_tweets.id.author_id&ids=123%2C456"
+          )
+        }
+        .respondWithRated("/twitter/rest/v2/tweets/tweetlookup/tweets.json")
+        .await
+      result.rate_limit === rateLimit
+      result.data === loadJsonAs[TweetsResponse]("/twitter/rest/v2/tweets/tweetlookup/tweets.json") // TODO setup fixtures
```
Comment (DanielaSfregola) [inline src/test/scala/com/danielasfregola/twitter4s/http/clients/rest/v2/tweets/TwitterV2TweetLookupClientSpec.scala:52]: I believe this TODO has been done :)

**Comment by DanielaSfregola** [comment]
So far is looking really good! 

There is some duplication, but it is well marked and, tbh, it will be very difficult to avoid -- so I am ok with it.

I look forward to see how this PR will evolve...so far, it looks very good! :)

File: src/test/scala/com/danielasfregola/twitter4s/http/clients/rest/v2/tweets/TwitterV2TweetLookupClientSpec.scala (hunk 1 above)
Comment (ConnorSinnott) [reply src/test/scala/com/danielasfregola/twitter4s/http/clients/rest/v2/tweets/TwitterV2TweetLookupClientSpec.scala:52]: Hey @DanielaSfregola! Thanks for popping in! Actually, would you mind explaining the intention behind the `/twitter/rest/**/*.json` and `/fixutres/rest/**/*.json` files? E.g. [twitter/user.json](https://github.com/DanielaSfregola/twitter4s/blob/24c8efae5ec2fc7ea6186575b386cef7b4b03352/src/test/resources/twitter/rest/users/user.json) and [fixtures/user.json](https://github.com/DanielaSfregola/twitter4s/blob/24c8efae5ec2fc7ea6186575b386cef7b4b03352/src/test/resources/fixtures/rest/users/user.json) appaer nearly identical. Would these files only differ if the structure of our deserailized model was different from what Twitter is serving?

Comment (DanielaSfregola) [reply src/test/scala/com/danielasfregola/twitter4s/http/clients/rest/v2/tweets/TwitterV2TweetLookupClientSpec.scala:52]: Correct: one is used to simulate the response from Twitter, the other asserts the information we parse (which may not necessarily be the same)

**Comment by ConnorSinnott** [comment]
Hey @DanielaSfregola! I'm marking this as review ready now that its been successfully capturing the tweets I've been running through it. Unfortunately Twitter's documentation doesn't clearly represent which properties are optional, and additionally I'm seeing some extra fields being returned by the live tests which aren't listed in the API docs, so I expect there will be a few more tweaks needed as we go. 

Since you last viewed:
- I've pulled the V2 endpoints into its own "V2" REST client. I'm thinking this would be a more future-proof approach since both versions serve similar data under similar method names.

- `Instant`s coming from Twitter now assume the more traditional form of `2021-06-11T18:30:12.000Z`, so the override of `EEE MMM dd HH:mm:ss ZZZZ yyyy` was colliding. I've changed the parsing of Instants to lazily iterate through both the old parser and `Instant.parse` and return the first one to succeed. 

- I've expanded the mock data to better represent a fully materialized Tweet. I was running into a few false-negatives from the test suite when the JSON parser wasn't configured right. If the same JSON parser is responsible for materializing the REST results as well as materializing the "expected" result, an error in the parser would cause both objects to be equally erroneous and the test will pass. I was thinking of checking equality against hand-written case classes, but I wanted to run that by you first.

**Comment by DanielaSfregola** [comment]
Writing the case classes by hand is certainly better -- the only reason why I didn't do it at the time is that I was developing 50+ endpoints and it wasn't practical for particularly verbose endpoints.

File: src/main/scala/com/danielasfregola/twitter4s/http/serializers/CustomFormats.scala [hunk 2]
```diff
@@ -18,9 +18,9 @@ private[twitter4s] object CustomFormats extends FormatsComposer {
 private[twitter4s] case object InstantSerializer
     extends CustomSerializer[Instant](format =>
       ({
-        case JInt(i)                                            => DateTimeFormatter.parseInstant(i.toLong)
-        case JLong(l)                                           => DateTimeFormatter.parseInstant(l)
-        case JString(s) if DateTimeFormatter.canParseInstant(s) => DateTimeFormatter.parseInstant(s)
+        case JInt(i)                                                         => DateTimeFormatter.parseInstant(i.toLong)
+        case JLong(l)                                                        => DateTimeFormatter.parseInstant(l)
+        case JString(s) if DateTimeFormatter.parseInstant(s).isDefined       => DateTimeFormatter.parseInstant(s).get
```
Comment (DanielaSfregola) [inline src/main/scala/com/danielasfregola/twitter4s/http/serializers/CustomFormats.scala:23]: Not a fan of `Option.get` ...let me see if we can keep the original signature!

File: src/main/scala/com/danielasfregola/twitter4s/http/serializers/JsonSupport.scala [hunk 3]
```diff
@@ -30,21 +30,25 @@ private[twitter4s] trait JsonSupport extends Json4sSupport {
...
   val builder: DateTimeFormatterBuilder = new DateTimeFormatterBuilder()
     .appendPattern("[EEE MMM dd HH:mm:ss Z yyyy]")
     .appendPattern("[yyyy-MM-dd HH:mm:ss Z]")
 
   val dateTimeFormatter = builder.toFormatter(locale)
 
-  def canParseInstant(s: String): Boolean = Try(parseInstant(s)).isSuccess
-
-  def parseInstant(s: String): Instant = formatter.parse(s).toInstant
+  def parseInstant(s: String): Option[Instant] = Seq(
```
Comment (DanielaSfregola) [inline src/main/scala/com/danielasfregola/twitter4s/http/serializers/JsonSupport.scala:48]: What about keeping the same signature def `parseInstant(s: String): Instant` and just do:

```
Try(v1InstanceFormatter(s)).getOrElse(v2InstantFormatter(s))
```

File: src/main/scala/com/danielasfregola/twitter4s/http/serializers/CustomFormats.scala (hunk 2 above)
Comment (DanielaSfregola) [reply src/main/scala/com/danielasfregola/twitter4s/http/serializers/CustomFormats.scala:23]: Yes, I believe you can avoid the `parseInstant(s).isDefined` and  `parseInstant(s).get` here if you keep the same signature as suggested in the comment

**Review by DanielaSfregola — CHANGES_REQUESTED** [review]
Overall looks very good! Can you please keep the same signature for the instant parser? I suggested you an equivalent simpler implementation for it :)

Cheers,
Daniela

File: src/main/scala/com/danielasfregola/twitter4s/http/serializers/JsonSupport.scala (hunk 3 above)
Comment (ConnorSinnott) [reply src/main/scala/com/danielasfregola/twitter4s/http/serializers/JsonSupport.scala:48]: Oh neat, I didn't realize you could `.getOrElse` on Try.

**Comment by ConnorSinnott** [comment]
Thanks for the speedy reviews! I've changed the signature of `parseInstant` back based on the `Try.getOrElse` approach suggested (neat). Additionally, I spent a few minutes creating the case class representation of these tweet responses and found that coordinates weren't being deserialized correctly 😅. So I've popped in a new custom serializer for `Tuple2[Double, Double]`. 

One thing to note is that I've placed the scala fixtures into a `fixtures` package within their relative tests.
```
- src/test/scala/com/danielasfregola/twitter4s/http/clients/rest/v2
  - tweets
    - fixtures
      - TweetResponseFixture.scala
      - TweetsResponseFixture.scala
    - TwitterTweetLookupClientSpec.scala
```
This is a bit of a departure from your structure, but the root `fixtures` directory isn't setup for scala. Let me know if I should try something else.

**Comment by ConnorSinnott** [comment]
I threw in one more last-minute change: The `errors` field of `TweetsResponse` and `TweetsResponse` are now `Seq[Error]` instead of `Option[Seq[Error]]`. The thinking is that other parts of this response are optional because they must be requested. The errors field isn't optional and will always be returned if they exist. So this change reflects that intention in the code.

**Review by DanielaSfregola — APPROVED** [review]

**Pull request merged**